python/triton/ops/matmul.py

```diff
@@ -155,33 +156,41 @@ def _call(a, b, dot_out_dtype, allow_tf32, fp8_fast_accum):
         assert a.shape[1] == b.shape[0], "incompatible dimensions"
         M, K = a.shape
         _, N = b.shape
+
+        # common type between a and b
+        ab_dtype = get_higher_dtype(a.dtype if a.dtype not in [tl.float8e4nv, tl.float8e4b15, tl.float8e5] else torch.float16,
+                                    b.dtype if b.dtype not in [tl.float8e4nv, tl.float8e4b15, tl.float8e5] else torch.float16)
+
         # allocates output
-        if a.dtype in [tl.float8e4nv, tl.float8e4b15, tl.float8e5] or\
-           b.dtype in [tl.float8e4nv, tl.float8e4b15, tl.float8e5]:
-            c_dtype = torch.float16
-        elif a.dtype in [torch.int8] or b.dtype in [torch.int8]:
-            c_dtype = torch.int32
-        else:
-            c_dtype = get_higher_dtype(a.dtype, b.dtype)
+        if (c_dtype is None):
+            c_dtype = ab_dtype
+
         c = torch.empty((M, N), device=device, dtype=c_dtype)
+
+        torch_tl_type_converter = {torch.float16: tl.float16, torch.bfloat16: tl.bfloat16, torch.float32: tl.float32, torch.int8: tl.int8, torch.int32: tl.int32}
+
+        # Allowed types for dot_out_type given the types of a and b.
+        type_preference_list = {}
+        type_preference_list[torch.float16] = [torch.float32, torch.float16]
+        type_preference_list[torch.bfloat16] = [torch.float32, torch.bfloat16]
+        type_preference_list[torch.float32] = [torch.float32]
+        type_preference_list[torch.int8] = [torch.int32]
         if dot_out_dtype is None:
-            if c_dtype in [torch.float16, torch.float32, torch.bfloat16]:
-                dot_out_dtype = tl.float32
-            else:
-                dot_out_dtype = tl.int32
+            dot_out_dtype = type_preference_list[c_dtype][0]
         else:
             assert isinstance(dot_out_dtype, torch.dtype), "dot_out_dtype must be a torch.dtype"
-            if dot_out_dtype == torch.float16:
-                dot_out_dtype = tl.float16
-            elif dot_out_dtype in [torch.float32, torch.bfloat16]:
-                dot_out_dtype = tl.float32
-            else:
-                dot_out_dtype = tl.int32
-        ab_dtype = True
+            assert dot_out_dtype in type_preference_list[a.dtype], "dot_out_dtype not compatible with the type of a"
+            assert dot_out_dtype in type_preference_list[b.dtype], "dot_out_dtype not compatible with the type of b"
+
+        dot_out_dtype = torch_tl_type_converter[dot_out_dtype]
+        ab_dtype = torch_tl_type_converter[ab_dtype]
+        c_dtype = torch_tl_type_converter[c_dtype]
+
+        # In this cases, there is no need to manually convert types to ab_dtype.
```

Comment:
Clarify why we do not need to convert instead of just saying what the code does, e.g.:

```python
# Tensor cores support input with mixed float8 types.
```

Reply:
good suggestion!